File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/Greedy/Knapsack.swift
import Foundation

class Items : Comparable {
    var wt : Int
    var cost : Int
    var density : Double

    init(_ w : Int, _ v : Int) {
        self.wt = w
        self.cost = v
        self.density = Double(self.cost) / Double(self.wt)
    }

    static func < (lhs: Items, rhs: Items) -> Bool {
        return lhs.density < rhs.density
    }

    static func > (lhs: Items, rhs: Items) -> Bool {
        return lhs.density > rhs.density
    }

    static func == (lhs: Items, rhs: Items) -> Bool {
        return lhs.density == rhs.density
    }
}

// Approximate solution.
func knapsackMaxCostGreedy(_ wt : inout [Int], _ cost : inout [Int], _ cap : Int) -> Int {
    var totalCost : Int = 0
    let n : Int = wt.count
    var capacity = cap
    
    var itemList : [Items] = []
    var i : Int = 0
    while (i < n) {
        itemList.append(Items(wt[i], cost[i]))
        i += 1
    }
    
    itemList.sort( by: >)// Decreasing order.

    i = 0
    while (i < n && capacity > 0) {
        if (capacity - itemList[i].wt >= 0) {
            capacity -= itemList[i].wt
            totalCost += itemList[i].cost
        }
        i += 1
    }
    return totalCost
}

// Testing code.
var wt : [Int] = [10, 40, 20, 30]
var cost : [Int] = [60, 40, 90, 120]
let capacity : Int = 50
let maxCost : Int = knapsackMaxCostGreedy( &wt, &cost,capacity)
print("Maximum cost obtained = " + String(maxCost))

/*
Maximum cost obtained = 150
*/

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/Greedy/Knapsack.ts
class Item {
	wt: number;
	cost: number;
	density: number;

	constructor(w: number, v: number) {
		this.wt = w;
		this.cost = v;
		this.density = v / w;
	}
}

function knapsack01Greedy(wt: number[], cost: number[], capacity: number): number {
	let totalCost = 0;
	const n = wt.length;
	const itemList: Item[] = [];

	for (let i = 0; i < n; i++) {
		itemList.push(new Item(wt[i], cost[i]));
	}

	itemList.sort((a, b) => b.density - a.density);

	for (let i = 0; i < n && capacity > 0; i++) {
		if (capacity - itemList[i].wt >= 0) {
			capacity -= itemList[i].wt;
			totalCost += itemList[i].cost;
		}
	}

	return totalCost;
}

/* Testing code */
const wt: number[] = [10, 40, 20, 30];
const cost: number[] = [60, 40, 90, 120];
const capacity: number = 50;
const maxCost: number = knapsack01Greedy(wt, cost, capacity);
console.log("Maximum cost obtained:", maxCost);

/*
Maximum cost obtained: 150
*/
